lib/GHCup/Types.hs

```diff
@@ -229,6 +229,7 @@ data Platform = Linux LinuxDistro
               | Darwin
               -- ^ must exit
               | FreeBSD
+              | OpenBSD
```

Comment:
This is a breaking change and will require us to create a new metadata file/version. So it's good to aggregate breaking changes, as in: add `NetBSD` and any other BSD you can think of.

Reply:
Why is it a breaking change (beyond having to change all the places where `Platform` type is used, which is done in this PR)? Aren't the metadata files just YAML (last time I checked)?

---

Off the top of my head, I can think of NetBSD, PC-BSD, DragonflyBSD, helloSystem; not sure of the status of PC-BSD or helloSystem (which is probably still in early stages, as that one's quite ambitious).

Apart from FreeBSD, though, OpenBSD is probably the next most popular one (or may even be on par, as it's known for having developers that actually use it day-to-day instead of a macOS system).

Anyway, adding these other systems will require some more research. I'm sure they'll package or pre-install `xdg-open`, it's unlikely they have their own homerolled solution to that, but what about, e.g. version checking? `uname -r`, or do they have their own special one like FreeBSD, especially those that forked off from it?

---

If we want to expand support as widely as possible, the current approach seems untenable. More troublingly, the idea that any new Platform requires backwards-incompatible changes to the metadata file format, but even the idea of grouping Linux distros under one `Platform` and having a `Distro` type seems quite arbitrary.

How much more similar to each other are Linux distros compared w/ BSD variants to each other? How about grouping BSDs together, or macOS as a BSD, or all POSIXy systems under a POSIX `Platform`? What about formerly Linux distros that now offer the option of running with a BSD kernel but same userland, header locations, package manager, etc.?

We should think about having sensible defaults to standards like POSIX or non-standard but widespread things like `xdg-open`, then for anything outside of that, delegating to external adapters which adhere to a formal API for extending GHCup with support for other platforms.

If we wanted to, we could hardcode support for anything with enough popularity/maintainers, but ideally, even officially supported platforms would just use external but first-party adapters. Hell, even the fallback behaviour could just be an adapter, although that seems unnecessary.

Reply:
@habibalamin this is breaking change because of derived instance for `Platform` will break for people using older `ghcup`, when the metadata is updated to contain "openbsd". (see `ToJSONKey Platform`)

So unless the addition of openbsd happens in `LinuxDistros` as "Linux_OpenBSD", any other change would require a metadata version bump.

So I suggest that just like `Linux LinuxDistro`, a new `BSD BSDDistro` is created to capture all the flavours. And the metadata migration could be simply "FreeBSD:" to "BSD_FreeBSD:" conversion.

Reply:
> 
> So I suggest that just like `Linux LinuxDistro`, a new `BSD BSDDistro` is created to capture all the flavours. And the metadata migration could be simply "FreeBSD:" to "BSD_FreeBSD:" conversion.

From what I know the BSDs are not really binary compatible and even run different kernels. So I'm not sure how we can call those "distros". I know there exist FreeBSD distros, but calling FreeBSD a BSD distro itself seems like a stretch. OpenBSD and FreeBSD are entirely separate operating systems, no?

Reply:
How many operating systems are out there that we will potentially support?

For linux distros it's a large number, so a catch-all there makes sense. If we don't have proper understanding of the operating system itself, how can we run on it?

Reply:
@hasufell ehh, many Linux distros run varying builds of the Linux kernel, but I think your point stands, especially as binary executables are not compatible; in fact, OpenBSD breaks backward compatibility with every release.

(In fact, I'm noting as I try to cross-compile GHC for AArch64 that its platform triple includes the version number by default, e.g. `x86_64-unknown-openbsd76` (which autotools `configure` seems to massage to remove the version when passing to the compiler and linker).)

The thing is, DragonflyBSD may retain compatibility with whatever release of FreeBSD it follows? I don't know how they all work, how incompatible they all are.

Either way, I don't think we've settled on the best abstraction for handling different platforms, but it's not an easy part to solve. But that's a topic for another day. For now, I'll get back to you when I can on the other BSDs and improving this PR to handle them as well.

Reply:
It doesn't matter much how we structure the types, in fact. We can keep json compatibility even if we unflatten them.

Reply:
>It doesn't matter much how we structure the types, in fact. We can keep json compatibility even if we unflatten them.

I'm not sure I understand what you're saying here or what you're responding to. Can you clarify?

Reply:
> It doesn't matter much how we structure the types, in fact. We can keep json compatibility even if we unflatten them.

means we can write a custom JSON instance for `Platform` in the new version of ghcup, such that there is not need for nesting as done now with `Linux`. So adding "OpenBSD" and "NetBSD" at the level of `Platform` is fine.

I am ok with this idea, this allows maximum flexibility going forward. And I don't have very clear understanding on the merits of flattened vs nested, other than the "_How many operating systems are out there that we will potentially support?_" isn't going to be too many, so flattened is simpler in this case.

Reply:
`ghcup` should at least be able to read the metadata containing newer OS fields. Whether it can handle that correctly should not be an expectation. I think it is fair to assume users to use latest `ghcup` if they are on this newly supported OS.

Is this understanding correct?

Reply:
> older ghcup should at least be able to read the metadata containing newer OS fields. Whether it can handle that correctly should not be an expectation

Yes, forward-compatibility is desirable. For the platforms, we use the `FromJSONKey` parser:

https://github.com/haskell/ghcup-hs/blob/9d6f101435ce1ea1ead13f92cbf283a7f391106b/lib/GHCup/Types/JSON.hs#L148-L168

I'm not sure how to write a FromJSONKey parser that *ignores* keys it doesn't understand, without messing with the types on the Haskell side.

I'm not sure I want an `OtherPlatform` constructor.

Reply:
The way to ignore unknown keys would require custom instance for `Map`. Though we would want to do this only for reading metadata in ghcup, and use the usual instance for check of metadata.

`type PlatformReqSpec = MapWithIgnore Platform PlatformReqVersionSpec`

It may be good to use this for arch as well. as the `wasm32` or `js*` will arrive some time in future
`type ArchitectureSpec = MapWithIgnore Architecture PlatformSpec`

Reply:
> The way to ignore unknown keys would require custom instance for Map. Though we would want to do this only for reading metadata in ghcup, and use the usual instance for check of metadata.

You mean a newtype for Map? Or a custom overloaded instance?

Reply:
a newtype `MapIgnoreUnknownKeys`

which  will only be used for reading yaml, then it will be converted to Map, and rest of code remains same

Reply:
Sounds ok

Reply:
https://github.com/haskell/ghcup-hs/pull/1144 has been merged